stl/inc/coroutine

```diff
@@ -0,0 +1,337 @@
+//   Returns: A coroutine handle `h` that refers to the coroutine whose promise object is referenced by `p`. [Note:
+//            `addressof(h.promise()) == addressof(p)` is consequently `true`.-end note]
+//   Throws: Nothing.
+// * coroutine_handle<>::address:
+//   Returns: A non-null pointer value if and only if `*this` refers to a coroutine.
+// * coroutine_handle<>::from_address and coroutine_handle<Promise>::from_address:
+//   Expects: ...
+//   Returns: A coroutine handle that refers to the same coroutine, if any, as the coroutine handle from which `addr`
+//            was obtained. [Note: For a coroutine handle `h`, `from_address(h.address()) == h` is `true`.]
+// * coroutine_handle<>::done: s/Requires/Expects/
+// * coroutine_handle<>::operator() and coroutine_handle<>::resume:
+//   Expects: `*this` refers to a coroutine suspended at other than its final suspend point.
+//   Effects: Resumes the execution of the coroutine.
+// * coroutine_handle<>::destroy: s/Requires/Expects/
+// * coroutine_handle<Promise>::promise: s/Requires/Expects/
+
+// intrinsics used in the implementation of coroutine_handle
+_EXTERN_C
+size_t _coro_resume(void*); // FIXME: Calling convention?
+#pragma intrinsic(_coro_resume) // FIXME: Is `#pragma intrinsic` necessary?
+
+void _coro_destroy(void*); // FIXME: Calling convention?
+#pragma intrinsic(_coro_destroy) // FIXME: Is `#pragma intrinsic` necessary?
+_END_EXTERN_C
+
+_STD_BEGIN
+
+// STRUCT TEMPLATE coroutine_traits
+template <class _Ret, class = void>
+struct _Coroutine_traits {};
+
+template <class _Ret>
+struct _Coroutine_traits<_Ret, void_t<typename _Ret::promise_type>> {
+    using promise_type = typename _Ret::promise_type;
+};
+
+template <class _Ret, class...>
+struct coroutine_traits : _Coroutine_traits<_Ret> {};
+
+// STRUCT TEMPLATE coroutine_handle
+template <class = void>
+struct coroutine_handle;
+
+struct _Resumable_frame {
+    using _Fn = void __cdecl(void*);
+
+    _Fn* _Resume;
+    unsigned short _Index;
+    unsigned short _Flags;
+};
+
+inline constexpr intptr_t _Noop_sentinel_value = -1;
+
+template <>
+struct coroutine_handle<void> {
+    constexpr coroutine_handle() noexcept = default;
+    constexpr coroutine_handle(nullptr_t) noexcept {}
+
+    coroutine_handle& operator=(nullptr_t) noexcept {
+        _Ptr = nullptr;
+        return *this;
+    }
+
+    _NODISCARD constexpr void* address() const noexcept {
+        return _Ptr;
+    }
+
+    _NODISCARD static constexpr coroutine_handle from_address(void* const _Addr) noexcept { // strengthened
+        coroutine_handle _Result;
+        _Result._Ptr = _Addr;
+        return _Result;
+    }
+
+    constexpr explicit operator bool() const noexcept {
+        return _Ptr != nullptr;
+    }
+
+    _NODISCARD bool done() const noexcept { // strengthened
+        if (reinterpret_cast<intptr_t>(_Ptr) == _Noop_sentinel_value) {
+            return false;
+        }
+
+        // _STL_ASSERT(_Ptr != nullptr);

```

Comment:
Reviewer note: All of these `_STL_ASSERT`s are commented out because I intended this to be a `(core)` header. Our headers marked `(core)` don't require or cause linking with the library. Consequently, the assertion failure handler might not available.

I suppose with today's technology we could uncomment these and make them `_STL_INTERNAL_CHECK`s.